core/src/main/scala/akka/persistence/cassandra/journal/CassandraRecovery.scala

```diff
@@ -20,21 +25,50 @@ trait CassandraRecovery extends ActorLogging {
     fromSequenceNr: Long,
     toSequenceNr:   Long,
     max:            Long
-  )(replayCallback: (PersistentRepr) => Unit): Future[Unit] =
-    queries
-      .eventsByPersistenceId(
-        persistenceId,
-        fromSequenceNr,
-        toSequenceNr,
-        max,
-        replayMaxResultSize,
-        None,
-        "asyncReplayMessages",
-        someReadConsistency,
-        someReadRetryPolicy
-      )
+  )(replayCallback: (PersistentRepr) => Unit): Future[Unit] = {
+
+    // No other writes for this pid should be taking place during recovery
+    // Expecting the result set to be small (< 10 rows)so call to all() should be safe
+    val tagProgress: Future[Map[String, TagProgress]] = preparedSelectTagProgressForPersistenceId.map(_.bind(persistenceId))
+      .flatMap(session.selectResultSet)
+      .map(rs => rs.all().foldLeft(Map.empty[String, TagProgress]) { (acc, row) =>
+        acc + (row.getString("tag") -> TagProgress(persistenceId, row.getLong("sequence_nr"), row.getLong("tag_pid_sequence_nr")))
+      })
+
+    Source.fromFutureSource(
+      tagProgress.map(tp => {
+        log.debug("Starting recovery with tag progress: {}", tp)
+        queries
+          .eventsByPersistenceId(
+            persistenceId,
+            fromSequenceNr,
```

Comment:
I think the `fromSequenceNr` is the from the latest snapshot. We must also find un-tagged events before snapshot. The starting point must be min(fromSeqNr, tagProgress.minSeqNr) and then filter...

Reply:
If events are tagged infrequently this could result in a large amount of scanning for each recovery even if the user has snapshotted. If there's no progress should we just start from the `fromSequenceNr` ?

Reply:
We now check the earliest sequenceNr and go from there until fromSequenceNr in a mode that just sends tag writes. Then does normal recovery starting at fromSequenceNr